ambry-router/src/main/java/com.github.ambry.router/SingleKeyManagementService.java

```diff
@@ -27,17 +27,24 @@
 
 /**
  * Single {@link KeyManagementService} which returns a default key {@link SecretKeySpec} for any
- * {@link #getKey(short, short)} calls.
+ * {@link #getKey(short, short)} calls but a different random key for {@link #getRandomKey()} calls.
  */
 public class SingleKeyManagementService implements KeyManagementService<SecretKeySpec> {
   private final SecretKeySpec secretKeySpec;
-  private final KMSConfig config;
   private volatile boolean enabled = true;
+  private final KeyGenerator keyGen;
+  private final String keyGenAlgo;
 
-  SingleKeyManagementService(KMSConfig config, String defaultKey) {
-    byte[] key = Hex.decode(defaultKey);
-    secretKeySpec = new SecretKeySpec(key, config.kmsKeyGenAlgo);
-    this.config = config;
+  SingleKeyManagementService(KMSConfig config, String defaultKey) throws GeneralSecurityException {
+    try {
+      byte[] key = Hex.decode(defaultKey);
+      keyGenAlgo = config.kmsKeyGenAlgo;
+      secretKeySpec = new SecretKeySpec(key, keyGenAlgo);
+      keyGen = KeyGenerator.getInstance(keyGenAlgo);
+      keyGen.init(config.kmsRandomKeySizeInBits);
+    } catch (NoSuchAlgorithmException e) {
```

Comment:
[`NoSuchAlgorithmException`](https://docs.oracle.com/javase/8/docs/api/java/security/NoSuchAlgorithmException.html) is a type of `GeneralSecurityException`.